/**
 * Copyright 2020 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";
package mindspore.ps.core;
option optimize_for = LITE_RUNTIME;

enum NodeCommand {
  TERMINATE = 0;
  REGISTER = 1;
  HEARTBEAT = 2;
  SEND_DATA = 3;
  // The worker or server asks the scheduler for metadata
  FETCH_METADATA = 4;
  FINISH = 5;
  COLLECTIVE_SEND_DATA = 6;
  // The scheduler actively sends metadata to the worker and server
  SEND_METADATA = 7;
  // This command is used to start scale out
  SCALE_OUT = 8;
  // This command is used to start scale in
  SCALE_IN = 9;
  // This command is used to synchronize the scale out status of the cluster
  SCALE_OUT_DONE = 10;
  // This command is used to synchronize the scale in status of the cluster
  SCALE_IN_DONE = 11;
  // This command is used to send user defined event.
  SEND_EVENT = 12;
}

enum NodeRole {
  SERVER = 0;
  WORKER = 1;
  SCHEDULER = 2;
}

message MessageMeta {
  // the command of this message,for example: register,heartbeat,data
  NodeCommand cmd = 1;
  // the request id of this message
  uint64 request_id = 2;
  // the role of the current node: worker,server,scheduler
  NodeRole role = 3;
  // the current Node rank id,the worker node range is:[0,numOfWorker-1], the server node range is:[0, numOfServer-1]
  uint32 rank_id = 4;
  // User-defined commands
  int32 user_cmd = 5;
}

message RegisterMessage {
  // ip
  string ip = 1;
  // the port of this node
  uint32 port = 2;
  // the current Node unique id:0,1,2...
  string node_id = 3;
  // the role of the node: worker,server,scheduler
  NodeRole role = 4;
}

message RegisterRespMessage {
  string node_id = 1;
}

message HeartbeatMessage {
  // the current Node unique id:0,1,2...
  string node_id = 1;
}

enum NodeState {
  NODE_STARTING = 0;
  NODE_FINISH = 1;
  NODE_READY = 2;
}

enum ClusterState {
  ClUSTER_STARTING = 0;
  CLUSTER_READY = 1;
  CLUSTER_EXIT = 2;
  NODE_TIMEOUT = 3;
  CLUSTER_SCALE_OUT = 4;
  CLUSTER_SCALE_IN = 5;
  CLUSTER_NEW_INSTANCE = 6;
  CLUSTER_ENABLE_FLS = 7;
  CLUSTER_DISABLE_FLS = 8;
}

message HeartbeatRespMessage {
  ClusterState cluster_state = 1;
  repeated ServersMeta servers_meta = 2;
  bool is_worker_or_server0 = 3;
}

message FetchServersMessage {
  string node_id = 1;
}

message FetchServersRespMessage {
  repeated ServersMeta servers_meta = 1;
}

message ServersMeta {
  uint32 rank_id = 1;
  string ip = 2;
  int32 port = 3;
  bool is_alive = 4;
  NodeRole role = 5;
  string node_id = 6;
}

message SendMetadataMessage {
  repeated ServersMeta servers_meta = 1;
  // the current worker number.
  int32 worker_num = 2;
  // the current server number.
  int32 server_num = 3;
  // the current cluster state.
  ClusterState cluster_state = 4;
  // The rank id of the node that received this message.
  uint32 rank_id = 5;
}

message FinishMessage {
  // the current Node unique id:0,1,2...
  string node_id = 1;
}

message CommMessage {
  MessageMeta pb_meta = 1;
  bytes data = 2;
}

// The scheduler will broadcast the worker/server numbers after scale out to all nodes.
message ScaleOutMessage {
  // the worker number after scale out
  int32 worker_num = 1;
  // the server number after scale out
  int32 server_num = 2;
}

// The scheduler will broadcast the worker/server numbers after scale in to all nodes.
message ScaleInMessage {
  // the worker number after scale in.
  int32 worker_num = 1;
  // the server number after scale in.
  int32 server_num = 2;
  // Determine whether the current node is a scale in node.
  bool is_node_scale_in = 3;
}

// This message is sent to the scheduler to notify the completion of scale out
message ScaleOutDoneMessage {
  string node_id = 1;
}

// This message is sent to the scheduler to notify the completion of scale out
message ScaleInDoneMessage {
  string node_id = 1;
}

// This message is sent by the worker/server to the scheduler, and the scheduler is broadcast the event to all other nodes.
message EventMessage {
  uint32 event = 1;
  string node_id = 2;
}

// scheduler broadcasts the event to all other nodes through this message
message EventRespMessage {
  uint32 event = 1;
}

message ScaleInFinishMessage {
  bool is_all_nodes_registered = 1;
}
